File: anki_connect.py
from pathlib import Path
from typing import Any

import structlog
from aiohttp import ClientSession

log = structlog.get_logger()


def get_api_url() -> str:
    is_inside_docker = Path("/.dockerenv").exists()
    if is_inside_docker:
        api_url = "http://host.docker.internal:8765"
    else:
        api_url = "http://localhost:8765"

    log.info("get_api_url", is_inside_docker=is_inside_docker, api_url=api_url)
    return api_url


api_url = get_api_url()


def get_client_session() -> ClientSession:
    return ClientSession(raise_for_status=True)


async def anki_connect_health_check() -> bool:
    async with get_client_session() as session:
        response = await session.get(api_url, raise_for_status=False)
        await response.text()
    is_connected = response.status == 200
    log.info("anki_connect_health_check", is_connected=is_connected)
    return is_connected


async def anki_connect_call(action: str, **kwargs) -> Any:
    payload = {
        "action": action,
        "params": kwargs,
        "version": 6,
    }
    log.info("anki_connect_call", action=action)

    async with get_client_session() as session:
        response = await session.post(api_url, json=payload)
        data = await response.json()
    if data["error"]:
        raise RuntimeError(data["error"])

    return data["result"]

File: cli/anki_connect.ts
import { exists } from "$std/fs/mod.ts";

export async function getApiUrl(): Promise<string> {
  const isInsideDocker = await exists("/.dockerenv");
  const apiUrl = isInsideDocker
    ? "http://host.docker.internal:8765"
    : "http://localhost:8765";

  console.info("getApiUrl", { isInsideDocker, apiUrl });
  return apiUrl;
}

export async function healthCheck(): Promise<boolean> {
  const apiUrl = await getApiUrl();
  const response = await fetch(apiUrl);
  await response.text();
  const isConnected = response.status === 200;

  console.info("healthCheck", { isConnected });
  return isConnected;
}

export async function ankiConnectCallRpc(
  action: string,
  params: Record<string, any>,
): Promise<any> {
  const payload = {
    action,
    params,
    version: 6,
  };
  console.info("jsonRpcCall", { action });

  const apiUrl = await getApiUrl();
  const response = await fetch(apiUrl, {
    method: "POST",
    body: JSON.stringify(payload),
  });

  const data = await response.json();
  if (data.error) {
    throw new Error(data.error);
  }
  return data.result;
}
